Implement softmax with NNPACK

Knight-X opened the pull request:
Summary:
  Implement softmax calculation with nnpack package. This would
  accelerate computation in resource limited device.
  Add test case for softmax with NNPACK

Commit:
Implement softmax with NNPACK

  Summary:
  Implement softmax calculation with nnpack package. This would
  accelerate computation in resource limited device.
  Add test case for softmax with NNPACK

Yangqing commented:
@Maratyszcza would you mind taking a look?

Commit:
Fix the bug of channels of input.

  Summary:
  The channel of softmax_nnpack should be H * W * C and batch_size
  is N. If channels is setted to N * C * H * W and batch_size is 1,
  the result is wrong.
  add test case for softmax.

Maratyszcza commented on caffe2/contrib/nnpack/nnpack_ops.cc:
```diff
@@ -335,9 +335,38 @@ class NNPACKLeakyReluOp final : public LeakyReluOp<float, CPUContext> {
+    auto& X = Input(0);
+    const auto N = X.dim32(0), C = X.dim32(1), H = X.dim32(2), W = X.dim32(3);
+    auto* Y = Output(0);
+    auto channels = C * H * W;
```
This is actually not right - see https://github.com/caffe2/caffe2/blob/master/caffe2/operators/softmax_op.cc#L9-L13 for the right index calculation.

Knight-X replied:
Right, I will fix the error calculation.

Commit:
fix the index calculation error.

Summary:
Fix the index calculation N and D which is reviewed by Maratyszcza.
set the atol and rtol value of test case in leaky_relu and softmax
to 1e-4, which is same as relu testcase

orionr commented:
Thank you for your contribution! PyTorch and Caffe2 are now officially merged with the PyTorch 1.0 preview release. Please submit on https://github.com/pytorch/pytorch if this is still an issue over there.
